cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaAvailability.java

```diff
@@ -0,0 +1,198 @@
+        return canRollBroker(descriptions, podId);
+    }
+
+    private Future<Boolean> canRollBroker(Future<Collection<TopicDescription>> descriptions, int podId) {
+        Future<Set<TopicDescription>> topicsOnGivenBroker = descriptions
+                .compose(topicDescriptions -> {
+                    log.debug("Got {} topic descriptions", topicDescriptions.size());
+                    return Future.succeededFuture(groupTopicsByBroker(topicDescriptions, podId));
+                }).recover(error -> {
+                    log.warn(error);
+                    return Future.failedFuture(error);
+                });
+
+        // 4. Get topic configs (for those on $broker)
+        Future<Map<String, Config>> topicConfigsOnGivenBroker = topicsOnGivenBroker
+                .compose(td -> topicConfigs(td.stream().map(t -> t.name()).collect(Collectors.toSet())));
+
+        // 5. join
+        return topicConfigsOnGivenBroker.map(topicNameToConfig -> {
+            Collection<TopicDescription> tds = topicsOnGivenBroker.result();
+            boolean canRoll = tds.stream().noneMatch(
+                td -> wouldAffectAvailability(podId, topicNameToConfig, td));
+            if (!canRoll) {
+                log.debug("Restart pod {} would remove it from ISR, stalling producers with acks=all", podId);
+            }
+            return canRoll;
+        }).recover(error -> {
+            log.warn("Error determining whether it is safe to restart pod {}", podId, error);
+            return Future.failedFuture(error);
+        });
+    }
+
+    private boolean wouldAffectAvailability(int broker, Map<String, Config> nameToConfig, TopicDescription td) {
+        Config config = nameToConfig.get(td.name());
+        ConfigEntry minIsrConfig = config.get(TopicConfig.MIN_IN_SYNC_REPLICAS_CONFIG);
+        int minIsr;
+        if (minIsrConfig != null && minIsrConfig.value() != null) {
+            minIsr = parseInt(minIsrConfig.value());
+            log.debug("{} has {}={}.", td.name(), TopicConfig.MIN_IN_SYNC_REPLICAS_CONFIG, minIsr);
+        } else {
+            minIsr = -1;
+            log.debug("{} lacks {}.", td.name(), TopicConfig.MIN_IN_SYNC_REPLICAS_CONFIG);
+        }
+
+        for (TopicPartitionInfo pi : td.partitions()) {
+            List<Node> isr = pi.isr();
+            if (minIsr >= 0) {
+                if (isr.size() < minIsr && contains(pi.replicas(), broker)) {
+                    logIsrReplicas(td, pi, isr);
+                    log.info("{}/{} is already underreplicated (|ISR|={}, {}={}); broker {} has a replica, " +
+                                    "so should not be restarted right now (it might be first to catch up).",
+                            td.name(), pi.partition(), isr.size(), TopicConfig.MIN_IN_SYNC_REPLICAS_CONFIG, minIsr, broker);
+                    return true;
+                } else if (isr.size() == minIsr && contains(isr, broker)) {
```

Comment:
Do we need to cover situation when min.isr = replication.factor? E.g. topic with 3 replicas and min.isr=3? Or is that handled somewhere else?